### PR TITLE
Reuse removed vars in mangler

### PR DESCRIPTION
**Pull request by boopathi**
Duplicate of #284. Rebased with current master.

Changes added to master / Issues reported for master and propagated to this branch:

+ [x] Destructuring bug fix (#390)
+ [x] recrawl scope (#381)
+ [x] topLevel option (#210)
+ [x] evalScopes helper (#371)
+ [x] #394 
+ [x] #365, #383 - Class Declaration Binding
+ [x] #400 Accept array as blacklist
+ [x] #414 renaming binding identifier with outer-inner collisions in bindingIdentifiers

### CODE REVIEW

**Comment by kangax** [comment]
@boopathi this looks great. The only thing I don't like is the "tracker" name since it doesn't really explain what it is or what it tracks and it might make reading code that uses it harder. Wouldn't "Scope" or "ScopeChain" be a better name?

**Comment by boopathi** [comment]
It is not `Scope`. It is basically -

```js
type Tracker {
  references: Map<Scope, CountedSet<String> >
  bindings: Map<Scope, Map<String, Binding> >
}
```

It counts references and maintains bindings.

Maybe call it `ScopeTracker` ?

**Comment by kangax** [comment]
Sure, that would be better.

File: packages/babel-plugin-minify-mangle-names/src/index.js [hunk 1]
```diff
@@ -18,76 +26,225 @@ module.exports = ({ types: t, traverse }) => {
...
 
+          if (!binding) {
+            // ignore the globals as it's available via Babel's API
+            if (scope.hasGlobal(name)) return;
+
+            // Ignore the NamedExports as they should NOT be mangled
             if (
-              path.parentPath.isMemberExpression({ property: node }) ||
-              path.parentPath.isObjectProperty({ key: node })
+              path.parentPath.isExportSpecifier() &&
```
Comment (vigneshshanmugam) [inline packages/babel-plugin-minify-mangle-names/src/index.js]: what about import specifiers? are they safe?

File: packages/babel-plugin-minify-mangle-names/src/scope-tracker.js [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+   * and Binding's Scope
+   * @param {Scope} scope
+   * @param {Binding} binding
+   * @param {String} name
+   */
+  addReference(scope, binding, name) {
+    let parent = scope;
+    do {
+      this.references.get(parent).add(name);
+      if (!binding) throw new Error("How did global come here");
```
Comment (vigneshshanmugam) [inline packages/babel-plugin-minify-mangle-names/src/scope-tracker.js]: rename the error message properly?

File: packages/babel-plugin-minify-mangle-names/src/index.js (hunk 1 above)
Comment (boopathi) [inline packages/babel-plugin-minify-mangle-names/src/index.js]: yes

File: packages/babel-plugin-minify-mangle-names/src/scope-tracker.js (hunk 2 above)
Comment (boopathi) [inline packages/babel-plugin-minify-mangle-names/src/scope-tracker.js]: done.

**Pull request merged**